openvpn-wire-protocol.xml

```diff
@@ -1533,15 +1541,95 @@ struct datakeys {
+      </t>
+      <t>
+        When a client or server lack the support for generating data channel keys using TLS Key material
+        export and do not annoounce the support in IV_CIPHER, the server may reject a client. At the same
+        time a client that does not get signalled by the server by `key-derivation tls-ekm` or
+        `protocol-flags tls-ekm` to use TLS Key material can abort the connection.
+      </t>
+      <t>
+        If support for older server or clients is desired that do not support TLS EKM, client and
+        server can fall back to the old mechanism of using the
+        <xref target="openvpnprf"> OpenVPN data channel PRF</xref>.
+      </t>
+      </section>
+      <section anchor="openvpnprf" title="OpenVPN data channel PRF">
+        <t>
+        This is a deprecated way of deriving the data channel keys. It should be only implemented if
+        compatibility with older OpenVPN versions is required that do not support key derivation via
+        the RFC5705 key material exporter. It also requires the MD5 algorithm which often
+        modern crypto libraries do not readily support anymore.</t>
+
+        <t>This uses the <tt>key_random</tt>, <tt>random1</tt>, and <tt>random2</tt> from the key
+        exchange messages. These will be prefixed with client and server here to indicate from which
+        packet these field are coming. Also the session id of the control channel will be used here</t>
+
+        <t>
+        Here TLS_PRF(seed, secret) is the TLS PRF function according to RFC 2246 with
+        MD5 and SHA1 as used in TLS 1.0/1.1.
+
+        <figure>
+          <sourcecode>
+seed = "OpenVPN master secret" | client.random1 | server.random1
+secret = client.key_random
+master_secret - TLS_PRF(seed, secret)
 
-        Older clients use the mechanism described in the section OpenVPN data channel
-        PRF.
+key_seed = "OpenVPN key expansion " | client.random2 | server.random2
+datakeys = TLS_PRF(key_seed, key_seed)
+          </sourcecode>
+        </figure>
       </t>
+      </section>
     </section>
+    <section title="Data channel key rotation">
+      <t>
+        Whenever a reneogiation of a key has finshed, a new data channel key is generated.
+        This new key will consider the pending key.  A pending key is considered valid for
+        decryption of packets with that key-id but will not be used for encryption yet.
+      </t>
+      <t>
+        After a set time period the pending key is promoted to being the active key and
+        the active key is set to be the retired key (sometimes also referred to as lame
+        duck key). The retired key will be used only for decryption but no longer for
+        encryption.
+      </t>
+      <t>
+        In the default configuration of OpenVPN, the time period of promoting a key from
+        pending to active is calucated as
+        <artwork>
+          min(handshake_window, renegotiation_time/2)
```

Comment:
What if `renegotiation_time` is 1? :D Is this then half a second? I ask because I use `--reneg-sec 1` in a test setup, and I am curious if sub second values are used.

Reply:
Then the implementation becomes broken and directly uses the new key since 1/2 = 0 in integer math:

From ssl.c:

```
static int
auth_deferred_expire_window(const struct tls_options *o)
{
    int ret = o->handshake_window;
    const int r2 = o->renegotiate_seconds / 2;

    if (o->renegotiate_seconds && r2 < ret)
    {
        ret = r2;
    }
    return ret;
}
```

I didn't want to include this that are more implementation bugs in corner cases into the RFC.

Reply:
Ok, that makes sense. Thanks for the code snippet. I'll modify my test setup!